File: project_3_client_server_analytics/albums.py
# this script will generate a CSV with ratings for albums
import random

albums = {
    "Meshuggah": ["Contradictions Collapse", "Destroy Erase Improve", "Chaosphere", "Nothing", "Catch Thirtythree", "ObZen", "Koloss", "The Violent Sleep of Reason", "Immutable"],
    "Vildhjarta": ["Omnilash", "Masstaden", "Thousands of Evils", "Masstaden Under Vaten", "Dar Skogen Sjunger"],
    "Humanity's Last Breath": ["Humanity's Last Breath", "Abyssal", "Vlade", "Ashen"]
}

means = {
    "Meshuggah": [1, 4, 2, 5, 5, 5, 3, 4, 4],
    "Vildhjarta": [1, 3, 4, 5, 4],
    "Humanity's Last Breath": [4, 2, 5, 4]
}

COUNT = 10000

if __name__ == "__main__":
    print("band,album,rating")
    artists = list(albums.keys())
    for i in range(COUNT):
        artist = random.choice(artists)
        artist_albums = albums[artist]
        album = random.choice(artist_albums)
        mean = means[artist][artist_albums.index(album)]
        rating = int(random.normalvariate(mu=mean, sigma=2))
        if rating < 0:
            rating = 0

        print(f"{artist},{album},{rating}")

File: project_3_client_server_analytics/analytics_lib/tests/albums.rs
use analytics_lib::query::{Aggregation, Query};
use analytics_lib::{csv::read_input_csv_file, query::Condition, solution};
use analytics_lib::dataset::{ColumnType, Dataset, Row, Value};

#[test]
fn test_albums() {
    let dataset = read_input_csv_file("../albums.csv");

    let query = Query::new(
        Condition::Equal(String::from("band"), Value::String(String::from("Meshuggah"))),
        String::from("album"),
        Aggregation::Average(String::from("rating")),
    );

    let result = solution::compute_query_on_dataset(&dataset, &query);


    let mut expected = Dataset::new(vec![
        (String::from("album"), ColumnType::String),
        (String::from("Average(rating)"), ColumnType::Integer),
    ]);
    expected.add_row(Row::new(vec![Value::String("Contradictions Collapse".to_string()), Value::Integer(1)]));
    expected.add_row(Row::new(vec![Value::String("Destroy Erase Improve".to_string()), Value::Integer(3)]));
    expected.add_row(Row::new(vec![Value::String("Chaosphere".to_string()), Value::Integer(1)]));
    expected.add_row(Row::new(vec![Value::String("Nothing".to_string()), Value::Integer(4)]));
    expected.add_row(Row::new(vec![Value::String("Catch Thirtythree".to_string()), Value::Integer(4)]));
    expected.add_row(Row::new(vec![Value::String("ObZen".to_string()), Value::Integer(4)]));
    expected.add_row(Row::new(vec![Value::String("Koloss".to_string()), Value::Integer(2)]));
    expected.add_row(Row::new(vec![Value::String("The Violent Sleep of Reason".to_string()), Value::Integer(3)]));
    expected.add_row(Row::new(vec![Value::String("Immutable".to_string()), Value::Integer(3)]));
}